### DISCUSSION TITLE
A problem about alias `latest`

### DISCUSSION THREAD

**Discussion by calyle** [jdx/mise#3168] [Q&A]
I have the latest version of node 20.

```console
$ mise v
2024.11.26 linux-x64 (2024-11-23)

$ mise list
Tool    Version          Config Source              Requested
node    20.18.1          ~/.config/mise/config.toml 20
```

I'd like to use the latest version of node for my project.
The latest stable version of node this time is

```console
$ mise latest node
23.3.0
```

And I run this to set node version of my project:

```console
$ mise use node@latest
mise ~/demo/mise.toml tools: node@20.18.1

$ mise install
mise all runtimes are installed

$ mise list
Tool    Version             Config Source              Requested
node    20.18.1 (outdated)  ~/demo/mise.toml           latest
```

I think the output above is unreasonable, because if I want to use the latest versoin of node (v23.3.0), I have to upgrade node 20 to node 23 instead of installing the latest node.

**Comment by syhol** [comment]
I think this has surprised a few people, just 2 weeks ago we had https://github.com/jdx/mise/issues/3005

I bet many people expect `mise use ...@latest` to always fetch the latest remote version.

**Reply by calyle** [reply]
Well, Thanks for your reply. I have read this issue. To tell the truth, I am not satisfied with this answer. The final solution is also upgrading node 20 to node 30.

Sometimes `mise` seems to be able to identify what I want.

If I want to install the latest version, I can install it by `mise install node@latest`

If node 20 has a new latest version, it will remind me to update node 20 to the latest node 20, but it doesn't upgrade node 20 to node 23. That's right.

When I want to use the latest node, it also knows what I want, but `mise` mistakenly suggests me upgrade node 20 to the latest version instead of install the latest version directly. Besides, if I run `mise use node@latest --pin`, `mise` will mistakenly considers that I want latest version of node 20, not latest node 23. And I can't do @jdx's suggestion to run `mise up --bump`, which will upgrade all languages.

I think it seems to be a matter of logic.

And in our real world, many times we don't use the latest stable version. As a example, we always use  java 1.8, java 17 and java 21, not java 23, so we have to remain the lower versions.

**Reply by calyle** [reply]
Uh...some problems seem to to be gone in the latest mise (v2024.11.27). I can use `mise use node@latest` to get the latest version of node, although node 20 exists locally.

But there are still some problems. If local `mise.toml` has the following content:

```toml
[tools]
node = "latest"
```

`mise` doesn't remind me `missing: node@23.3.0`, but remind me that my node 20 is outdated and have to upgrade it to node 23.

```console
$ mise list
Tool   Version             Config Source              Requested
node   20.18.1 (outdated)  ~/demo/mise.toml           latest
```

I think aliases are good, and greatly simplify language version settings.

I create a alias `stable` to `latest`:

```console
$ mise alias set node stable latest
$ mise alias
tool  alias        version
node  stable       latest
```

Why I set this? Because some projects use `nvm` to manage node version and the stable version is latest stable version

But I get the same output when I set node version to stable.

```console
$ mise use node@stable
mise ~/demo/mise.toml tools: node@20.18.1
$ mise list
Tool   Version             Config Source              Requested
node   20.18.1 (outdated)  ~/demo/mise.toml           stable
```

`mise` also reminds me to upgrade node 20 to node 23.

In my opinion, Setting an alias is like setting a variable.

**Reply by jdx** [reply]
> And I can't do @jdx's suggestion to run mise up --bump, which will upgrade all languages.

`mise up --bump node`?

**Reply by calyle** [reply]
> > And I can't do @jdx's suggestion to run mise up --bump, which will upgrade all languages.
> 
> `mise up --bump node`?

I'm sorry, this does solve the problem. But I think you know my idea. I just don't understand why I have to upgrade node 20, instead of installing it simply, And then uninstall after use. If I use `mise up --bump node`, I can indeed get the latest version, but afterwards, I must reinstall it back and uninstall node 23. Maybe I'm slightly OCD. Maybe it's unnecessary for you and you need to spend time to do more important things. And this's just my little suggestion.
